File: leetgo/go/LCR 150.cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/solution.go
// Created by k at 2024/05/14 14:51
// leetgo: dev
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func decorateRecord(root *TreeNode) (ans [][]int) {
	if root == nil {
		return [][]int{}
	}
	queue := []*TreeNode{root}
	for len(queue) != 0 {
		ql := len(queue)
		var lans []int
		for i := 0; i < ql; i++ {
			lans = append(lans, queue[i].Val)
			if queue[i].Left != nil {
				queue = append(queue, queue[i].Left)
			}
			if queue[i].Right != nil {
				queue = append(queue, queue[i].Right)
			}
		}
		queue = queue[ql:]
		ans = append(ans, lans)
	}

	return
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	root := Deserialize[*TreeNode](ReadLine(stdin))
	ans := decorateRecord(root)

	fmt.Println("\noutput:", Serialize(ans))
}

File: leetgo/cpp/LCR 150.cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/solution.cpp
// Created by k at 2024/05/14 15:05
// leetgo: dev
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/

#include <bits/stdc++.h>
#include <vector>
#include "LC_IO.h"
using namespace std;

// @lc code=begin

class Solution {
  public:
    vector<vector<int>> decorateRecord(TreeNode *root)
    {
        vector<vector<int>> ans;
        if (root == nullptr) {
            return ans;
        }
        vector<TreeNode *> queue = {root};
        while (!queue.empty()) {
            int l = queue.size();
            vector<int> lans;
            for (int i = 0; i < l; i++) {
                lans.push_back(queue[i]->val);
                if (queue[i]->left != nullptr) {
                    queue.push_back(queue[i]->left);
                }
                if (queue[i]->right != nullptr) {
                    queue.push_back(queue[i]->right);
                }
            }
            ans.push_back(lans);
            queue.erase(queue.cbegin(), queue.cbegin() + l);
        }
        return ans;
    }
};

// @lc code=end

int main()
{
    ios_base::sync_with_stdio(false);
    stringstream out_stream;

    TreeNode *root;
    LeetCodeIO::scan(cin, root);

    Solution *obj = new Solution();
    auto res      = obj->decorateRecord(root);
    LeetCodeIO::print(out_stream, res);
    cout << "\noutput: " << out_stream.rdbuf() << endl;

    delete obj;
    return 0;
}
